### PR TITLE
Integrate JoinSplit verifier

### LINKED ISSUE #1835: Integrate async JoinSplit Groth16 verifier with transaction verifier
In https://github.com/ZcashFoundation/zebra/pull/1713 we integrated the Spend and Output Groth16 async verification, after loading the public Sapling verification params.

Similarly, we'll download the JoinSplit proofs in ticket #3037, and validate them in ticket #322.

Then the Sprout JoinSplit verifier needs to be integrated with the rest of the system and fully tested.

## Todo

- [ ] Integrate the JoinSplit groth16 async verifier instance with the transaction verifier using the aforementioned keys

### PR DESCRIPTION
**Pull request by conradoplg**
## Motivation

We need to validate JoinSplit Groth16 proofs in order to be a fully validating node.

### Specifications

https://zips.z.cash/protocol/protocol.pdf#joinsplitdesc

image

## Solution

Of the above rules:

- The first should be already be covered by the parser;
- The second is implemented by this PR;
- The third seemed to be missing and is also implemented by this PR;
- The fourth was already implemented.

While developing this PR I identified an issue where all-zeroes proofs (or any malformed proof) would make Zebra panic. This PR adds a test for that and handles it by propagating the error while reading the proof into `bellman`. Ideally we'd like to that while parsing the transaction, but that would require a lot of code changes, so it was postponed and tracked in https://github.com/ZcashFoundation/zebra/issues/3179

This PR changes existing tests which would no longer pass since they used dummy generated transaction. They were changed to use an existing transaction, which is modified as needed.

Closes https://github.com/ZcashFoundation/zebra/issues/1835

## Review

Need to merge https://github.com/ZcashFoundation/zebra/pull/3128 first.

@dconnolly and/or @upbqdn can review it.

### Reviewer Checklist

  - [x] Code implements Specs and Designs
  - [x] Tests for Expected Behaviour
  - [x] Tests for Errors

## Follow Up Work

Improve type safety in https://github.com/ZcashFoundation/zebra/issues/3179

### CODE REVIEW

**Review by upbqdn — DISMISSED** [review]
All looks good to me here.

File: zebra-consensus/src/transaction.rs [hunk 1]
```diff
@@ -314,6 +315,10 @@ where
                 check::non_coinbase_expiry_height(&req.height(), &tx)?;
             }
 
+            // Either v_{pub}^{old} or v_{pub}^{new} MUST be zero.
+            // https://zips.z.cash/protocol/protocol.pdf#joinsplitdesc
```
Comment (dconnolly) [inline zebra-consensus/src/transaction.rs]:
```suggestion
            // Consensus rule:
            //
            // > Either v_{pub}^{old} or v_{pub}^{new} MUST be zero.
            //
            // https://zips.z.cash/protocol/protocol.pdf#joinsplitdesc
```

File: zebra-consensus/src/primitives/groth16/tests.rs [hunk 2]
```diff
@@ -110,9 +112,21 @@ async fn verify_sapling_groth16() {
         .unwrap()
 }
 
+#[derive(Clone, Copy)]
+enum Groth16OutputModification {
+    ZeroCMU,
+    ZeroProof,
+}
+
```
Comment (dconnolly) [inline zebra-consensus/src/primitives/groth16/tests.rs:120]: 👍

File: zebra-consensus/src/transaction.rs [hunk 3]
```diff
@@ -461,7 +466,7 @@ where
         .and(Self::verify_sprout_shielded_data(
             joinsplit_data,
             &shielded_sighash,
-        ))
+        )?)
```
Comment (dconnolly) [inline zebra-consensus/src/transaction.rs:472]: 👀

File: zebra-consensus/src/transaction.rs [hunk 4]
```diff
@@ -632,16 +637,23 @@ where
...
-            // Then, pass those items to self.joinsplit to verify them.
+            for joinsplit in joinsplit_data.joinsplits() {
+                // Consensus rule: The proof π_ZKSpend MUST be valid given a
+                // primary input formed from the relevant other fields and h_{Sig}
+                //
+                // Queue the verification of the Groth16 spend proof
+                // for each JoinSplit description while adding the
+                // resulting future to our collection of async
+                // checks that (at a minimum) must pass for the
+                // transaction to verify.
```
Comment (dconnolly) [inline zebra-consensus/src/transaction.rs:655]:
```suggestion
                // Consensus rule: The proof π_ZKSpend MUST be valid given a
                // primary input formed from the relevant other fields and h_{Sig}
                //
                // Queue the verification of the Groth16 spend proof
                // for each JoinSplit description while adding the
                // resulting future to our collection of async
                // checks that (at a minimum) must pass for the
                // transaction to verify.
                //
                // https://zips.z.cash/protocol/protocol.pdf#joinsplitdesc
```

File: zebra-consensus/src/transaction.rs [hunk 5]
```diff
@@ -661,7 +673,7 @@ where
             checks.push(ed25519_verifier.oneshot(ed25519_item));
         }
 
-        checks
+        Ok(checks)
```
Comment (dconnolly) [inline zebra-consensus/src/transaction.rs:681]: 👍

File: zebra-consensus/src/transaction/check.rs [hunk 6]
```diff
@@ -164,6 +164,27 @@ pub fn output_cv_epk_not_small_order(output: &Output) -> Result<(), TransactionE
     }
 }
 
+/// Check if JoinSplits in the transaction have one of its v_{pub} values equal
+/// to zero.
+///
+/// https://zips.z.cash/protocol/protocol.pdf#joinsplitdesc
```
Comment (dconnolly) [inline zebra-consensus/src/transaction/check.rs]:
```suggestion
/// Check if JoinSplits in the transaction have one of its v_{pub} values equal
/// to zero.
///
/// <https://zips.z.cash/protocol/protocol.pdf#joinsplitdesc>
```

File: zebra-consensus/src/transaction/tests.rs [hunk 7]
```diff
@@ -1476,6 +1463,63 @@ fn mock_sprout_join_split_data() -> (JoinSplitData<Groth16Proof>, ed25519::Signi
...
+    joinsplit_data: &mut JoinSplitData<Groth16Proof>,
+    modification: JoinSplitModification,
+) {
+    match modification {
+        JoinSplitModification::CorruptSignature => {
+            let mut sig_bytes: [u8; 64] = joinsplit_data.sig.into();
+            // Flip a bit from an arbitrary byte of the signature.
+            sig_bytes[10] ^= 0x01;
+            joinsplit_data.sig = sig_bytes.into();
+        }
```
Comment (dconnolly) [inline zebra-consensus/src/transaction/tests.rs:1499]: 👍

File: zebra-consensus/src/transaction/tests.rs [hunk 8]
```diff
@@ -1476,6 +1463,63 @@ fn mock_sprout_join_split_data() -> (JoinSplitData<Groth16Proof>, ed25519::Signi
...
+            let joinsplit = joinsplit_data
+                .joinsplits_mut()
+                .next()
+                .expect("must have a JoinSplit");
+            {
+                // A proof is composed of three field elements, the first and last having 48 bytes.
+                // (The middle one has 96 bytes.) To corrupt the proof without making it malformed,
+                // simply swap those first and last elements.
+                let (first, rest) = joinsplit.zkproof.0.split_at_mut(48);
+                first.swap_with_slice(&mut rest[96..144]);
```
Comment (dconnolly) [inline zebra-consensus/src/transaction/tests.rs:1510]: Cool!

File: zebra-consensus/src/transaction/tests.rs [hunk 9]
```diff
@@ -1476,6 +1463,63 @@ fn mock_sprout_join_split_data() -> (JoinSplitData<Groth16Proof>, ed25519::Signi
...
+            }
+        }
+        JoinSplitModification::ZeroProof => {
+            let joinsplit = joinsplit_data
+                .joinsplits_mut()
+                .next()
+                .expect("must have a JoinSplit");
+            joinsplit.zkproof.0 = [0; 192];
```
Comment (dconnolly) [inline zebra-consensus/src/transaction/tests.rs:1518]: 👍

**Comment by dconnolly** [comment]
Some small rustdoc suggestions, kicked off a full sync with this branch, otherwise looking great 🥳 

https://github.com/ZcashFoundation/zebra/runs/4490896977

**Comment by conradoplg** [comment]
I've added @dconnolly 's suggestions so this needs re-approval. There is an issue with the deployed node whose details I posted over Slack, but I'm not sure if it's related to this PR or not

**Review by dconnolly — APPROVED** [review]

**Pull request merged**